### ISSUE TITLE
Mobile version

### ISSUE THREAD

**Issue by piclemx** [keeweb/keeweb#534]
Hi, 

I just wanna know if a mobile version will be create?

**Comment by antelle** [comment]
Maybe, but not now. A lot of work is required...

**Comment by piclemx** [comment]
What are the current work on this project?

**Comment by antelle** [comment]
On mobile? It's not being developed at all.

**Comment by alex-shpak** [comment]
I have prototype of cordova wrapper for application but it's still under early development.
Also there is some interface changes required.

**Comment by piclemx** [comment]
Very happy to hear that 😄

**Comment by piclemx** [comment]
@alex-shpak do you have a idea when this will be ready?

**Comment by alex-shpak** [comment]
Hi, didn't have chance to work on it recent weeks, hope to back on track soon.
Here is how it looks now:

![keemob](https://cloud.githubusercontent.com/assets/5515443/26451042/9bbb9fb6-415a-11e7-9898-d5ef2fcd5068.gif)

**Comment by piclemx** [comment]
Ok good 😄  ! Will work on iOS ?

**Comment by hultberg** [comment]
As my keepass database is synced via Dropbox, I use MiniKeePass on iOS while syncing via the official dropbox app.

**Comment by alex-shpak** [comment]
@piclemx I hope it will :)

**Comment by PacketPerception** [comment]
@alex-shpak I'd be happy to help out on iOS if you want to share the project.

**Comment by alex-shpak** [comment]
@PacketPerception 
I published repo here: https://github.com/alex-shpak/keemob

In order to run it you need to clone keeweb, uncomment cordova launcher [here](https://github.com/keeweb/keeweb/blob/master/app/scripts/comp/launcher.js#L6), build it and put `index.html` into `www` folder.

Then use `cordova build` and `cordova run android`

**Comment by noxarivis** [comment]
The downloaded index HTML works fine in chrome on android

**Comment by noxarivis** [comment]
How do I download files on mobile?

**Comment by alex-shpak** [comment]
you can download this file https://app.keeweb.info/index.html on desktop, put it into google drive and then download on android, or you can use directly that web page.

**Comment by noxarivis** [comment]
@alex-shpak I'm sure u misunderstood me

**Comment by alex-shpak** [comment]
Ah, you mean how to download file from keepass storage?
I think it's not possible in UI right now

**Comment by Maxiride** [comment]
@alex-shpak Nice to hear that a mobile app is somehow under development, are there any updates after almost an year?

**Comment by SilencerWeb** [comment]
@Maxiride, +1

**Comment by alex-shpak** [comment]
Hi!
Sorry for late response,

I have prototype for testing on my android phone, unfortunately there are some issues like:
1. UI is not scalable for mobile, it requires a lot of changes
2. Performance, file decryption is irritably slow
3. Missing clipboard feature is support (that one is fixable)
4. With current implementation (via content provider) merging files doesn't work if device was offline

And also waiting for https://github.com/keeweb/keeweb/pull/674

I guess there are two ways fixing UI: update existing project UI, which might be rather big amount of changes, or use https://github.com/keeweb/kdbxweb and implement separate mobile UI with it.

@antelle what do you think?

**Comment by antelle** [comment]
@alex-shpak yeah, (1.) is the main problem for now and it's unlikely to be solved. Nor do I see any reasonable approach to solve (4.) within the current codebase.
I'd like to share my thoughts about future plans, how to solve (2.), (3.), and (4.).
1. kdbxweb (the core library that deals with the format and encryption) will be rewritten in a native language, I don't know which one, it depends on what will compile to WebAssembly and native, probably C++, but maybe there will be something better, [the list is already huge](https://github.com/appcypher/awesome-wasm-langs): I would like to postpone it a bit, to wait until the ecosystem gets more mature
2. keeweb model (business logic, dealing with files, syncing, etc...) will be also migrated to WebAssembly
3. UI and views pretty much depend on trends in WebAssembly: if there's a reasonable UI framework, we can use it, or we may continue to use Electron with current UI implementation for some time until there's a better alternative
4. mobile UI, if not supported automatically by the framework, can be created using the same model, for mobile it can be less complicated (because less interface is required on mobile), so it could be easy to create using platform-native languages, with native look&feel, sync, performance, and other stuff users might expect of well-behaving apps

Things will remain as is for at least this fall: it's unlikely we can make a good decision now, when WASM has just appeared.
Sorry for such a long update 😌 I didn't have time to write a short letter (c) Mark Twain

**Comment by ConcurrentHashMap** [comment]
I am also looking for a good replacement of Keepass2Android (that is really great) for iOS. All the available apps don't look nice to me, very outdated and not really in active development.

What are the people using a KeePass database needing on their smartphone? I can just call from myself, but in 99,9% of the time, I open my KeePass database on mobile, I just need to enter my credentials on an app that forgot them or where I logged out remotely from web from all sessions. So it's basically read-only stuff. All the advanced stuff like merging files, ... isn't really a thing you need to do on mobile.

So I would like to discuss this issue a bit more, as I'm very interested in a workaround for using KeeWeb also on iOS (and as Keepass2Android is very good, stable and actively maintained I would suggest to really focus on iOS for now!)

@alex-shpak already made some good suggestions, but they still lack a good usability.

What is possible at the moment?
- I can access https://app.keeweb.info/index.html from Safari and add it to the iOS homescreen, therefore it looks like an app and it will also be run in fullscreen mode. The problem here: I don't want to trust this website, as it seems strange to me, opening a website for then entering my most sensitive information, although it is still secured by SSL (think of open wifi hotspots that might redirect DNS etc...).
- I could also download this file somewhere and then run it from lets say the iOS 11 File app. Not sure if this works 100%, I only tested this with a basic HTML file and seems to render HTML (but I also know that iOS limits JavaScript etc. on certain apps).

So my idea is: Isn't there a way of wrapping the index.html file inside a basic (system integrated Safari) webviewer and limit this app for not allowing any outgoing network connection?

For a first workaround I also found at good blogpost about the same thoughts here: http://mybyways.com/blog/trying-out-keeweb-and-minikeepass-password-manager

**Comment by alex-shpak** [comment]
@antelle Speaking of WASM, recently Golang 1.11 was released with WASM support, It is experimental so far, but might be a good candidate.

**Comment by dploeger** [comment]
To update this:

I just wrapped up a cordova project inside KeeWeb, started it on my iOS device (because the simulator somehow doesn't support WebAssembly 🙄) and actually only see minor problems (and some room for improvement):

* Direct access to files isn't supported (it's accepting a file from iCloud drive, but then it's uploading it to the app). This is the biggest problem IMHO, because we would need to support direct access. KyPass (the app I use) does this by directly supporting iCloud, which would possibly be the way to go.
* Some minor UI improvements (i.e. the "remove database"-X-symbol is reeeeally tiny 😉)
* Integration of Cordova into the build system
* (future:) Support for vendor-specific features like (can only tell about iOS here):
  * FaceID/TouchID (@alex-shpak seems to have gone some steps into that direction 👍)
  * Mobile browser Password-Integration (iOS has that since iOS 12 and it really is a great feature, because it allows you to select a password from KyPass on every password field)

But that would be it, I guess.

**Comment by alex-shpak** [comment]
So, I updated cordova wrapper to latest versions of cordova/keeweb
https://github.com/alex-shpak/keemob/releases/tag/0.0.4

Usual disclaimer: I use this app at my phone but it is very-very beta. Use at your own risk.
So far I didn't have troubles with it over a year.

**Comment by kegowo** [comment]
@alex-shpak Thanks for updating the app. The new version works well and I hope this gets merged into keeweb. We can't use your version because of the app name.

**Comment by antelle** [comment]
I won't be able to support the mobile version, so I'd say, supporting it this way, when an app is another project that wraps KeeWeb, is the way to go here.
Could you please explain your issue with the app name?

**Comment by kegowo** [comment]
Apps with names that look like malware get removed and blocked on phones managed by corporate security compliance. It's also a risk to cross an international border with an app that could be mistaken as a hacker program.

**Comment by kegowo** [comment]
I compile my own version of the app but if you need an example of a safe name `com.github.keeweb.mobile` would work.

Other password managers with dumb names are also blocked like the first two listed at https://itsfoss.com/password-managers-linux/.

**Comment by aisbergde** [comment]
above is a video containing fingerprint authentication. But it looks like this is not a general feature, but only a kind of demonstration. Are there plans to enable this for smartphones?

**Comment by antelle** [comment]
You're right, it's a demo made with Cordova, a mobile wrapper. Was never released afaik...
